File: microservices/api/api-scala/app/controllers/DataController.scala
package controllers

import javax.inject._
import play.api._
import play.api.mvc._
import play.api.libs.json._
import play.api.libs.ws._
import models._
import services.JwtService
import actions.AuthAction
import io.swagger.annotations._
import scala.concurrent.{ExecutionContext, Future}
import scala.concurrent.duration._

/**
 * Data controller for MindsDB integration
 */
@Api(value = "Data", description = "Data query operations")
@Singleton
class DataController @Inject()(
  val controllerComponents: ControllerComponents,
  authAction: AuthAction,
  ws: WSClient,
  config: Configuration
)(implicit ec: ExecutionContext) extends BaseController with Logging {

  private val mindsdbUrl = config.get[String]("app.mindsdb.url")
  private val mindsdbTimeout = config.get[Duration]("app.mindsdb.timeout")

  /**
   * Query MindsDB
   */
  @ApiOperation(
    value = "Query MindsDB",
    notes = "Execute SQL queries against MindsDB",
    httpMethod = "POST",
    response = classOf[QueryResponse],
    authorizations = Array(new Authorization("Bearer"))
  )
  @ApiImplicitParams(Array(
    new ApiImplicitParam(
      name = "body",
      value = "Query request",
      required = true,
      dataType = "models.QueryRequest",
      paramType = "body"
    )
  ))
  @ApiResponses(Array(
    new ApiResponse(code = 200, message = "Query results", response = classOf[QueryResponse]),
    new ApiResponse(code = 400, message = "Bad request", response = classOf[ErrorResponse]),
    new ApiResponse(code = 401, message = "Unauthorized", response = classOf[ErrorResponse]),
    new ApiResponse(code = 500, message = "Server error", response = classOf[ErrorResponse])
  ))
  def queryMindsDB = authAction.async(parse.json) { implicit request =>
    request.body.validate[QueryRequest] match {
      case JsError(errors) =>
        Future.successful(BadRequest(Json.toJson(
          ErrorResponse("Invalid request body", Some("VALIDATION_ERROR"), 400)
        )))
        
      case JsSuccess(queryRequest, _) =>
        val startTime = System.currentTimeMillis()
        val queryUrl = s"$mindsdbUrl/api/sql/query"
        
        ws.url(queryUrl)
          .withHttpHeaders(
            "Content-Type" -> "application/json",
            "Accept" -> "application/json"
          )
          .withRequestTimeout(queryRequest.timeout.map(_.seconds).getOrElse(mindsdbTimeout))
          .post(Json.obj("query" -> queryRequest.query))
          .map { response =>
            val executionTime = System.currentTimeMillis() - startTime
            
            response.status match {
              case 200 =>
                val data = response.json match {
                  case JsArray(arr) => JsArray(arr)
                  case obj: JsObject => JsArray(Seq(obj))
                  case _ => JsArray(Seq())
                }
                
                val queryResponse = QueryResponse(
                  data = data,
                  count = data.value.size,
                  executionTime = executionTime,
                  metadata = Some(Json.obj(
                    "query" -> queryRequest.query,
                    "timestamp" -> System.currentTimeMillis()
                  ))
                )
                Ok(Json.toJson(queryResponse))
                
              case 400 =>
                BadRequest(Json.toJson(
                  ErrorResponse(
                    "Invalid query syntax",
                    Some("QUERY_SYNTAX_ERROR"),
                    400,
                    Some(Json.obj("query" -> queryRequest.query))
                  )
                ))
                
              case _ =>
                InternalServerError(Json.toJson(
                  ErrorResponse(
                    "MindsDB query failed",
                    Some("MINDSDB_ERROR"),
                    500,
                    Some(Json.obj(
                      "status" -> response.status,
                      "body" -> response.body
                    ))
                  )
                ))
            }
          }
          .recover {
            case e: java.util.concurrent.TimeoutException =>
              logger.error("MindsDB query timeout", e)
              RequestTimeout(Json.toJson(
                ErrorResponse(
                  "Query timeout",
                  Some("QUERY_TIMEOUT"),
                  408,
                  Some(Json.obj("timeout" -> mindsdbTimeout.toString))
                )
              ))
              
            case e: Exception =>
              logger.error("MindsDB query error", e)
              InternalServerError(Json.toJson(
                ErrorResponse("Internal server error", Some("SERVER_ERROR"), 500)
              ))
          }
    }
  }
}

File: microservices/api/api-java/app/controllers/DataController.java
package controllers;

import actions.AuthRequest;
import com.auth0.jwt.interfaces.DecodedJWT;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.typesafe.config.Config;
import io.swagger.annotations.*;
import models.ErrorResponse;
import models.QueryRequest;
import models.QueryResponse;
import play.Logger;
import play.libs.Json;
import play.libs.ws.WSClient;
import play.mvc.Controller;
import play.mvc.Result;
import play.mvc.Http;
import javax.inject.Inject;
import javax.inject.Singleton;
import java.time.Duration;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

/**
 * Data controller for MindsDB integration
 */
@Api(value = "Data", description = "Data query operations")
@Singleton
public class DataController extends Controller {
    
    private static final Logger.ALogger logger = Logger.of(DataController.class);
    
    private final WSClient wsClient;
    private final String mindsdbUrl;
    private final Duration mindsdbTimeout;
    
    @Inject
    public DataController(WSClient wsClient, Config config) {
        this.wsClient = wsClient;
        this.mindsdbUrl = config.getString("mindsdb.url");
        this.mindsdbTimeout = config.getDuration("mindsdb.timeout");
    }
    
    /**
     * Query MindsDB
     */
    @ApiOperation(
        value = "Query MindsDB",
        notes = "Execute SQL queries against MindsDB",
        httpMethod = "POST",
        response = QueryResponse.class,
        authorizations = {@Authorization("Bearer")}
    )
    @ApiImplicitParams({
        @ApiImplicitParam(
            name = "body",
            value = "Query request",
            required = true,
            dataType = "models.QueryRequest",
            paramType = "body"
        )
    })
    @ApiResponses(value = {
        @ApiResponse(code = 200, message = "Query results", response = QueryResponse.class),
        @ApiResponse(code = 400, message = "Bad request", response = ErrorResponse.class),
        @ApiResponse(code = 401, message = "Unauthorized", response = ErrorResponse.class),
        @ApiResponse(code = 500, message = "Server error", response = ErrorResponse.class)
    })
    @AuthRequest
    public CompletionStage<Result> queryMindsDB(Http.Request request) {
        JsonNode json = request.body().asJson();
        if (json == null) {
            return CompletableFuture.completedFuture(
                badRequest(Json.toJson(new ErrorResponse("Missing request body", "MISSING_BODY", 400)))
            );
        }
        
        QueryRequest queryRequest;
        try {
            queryRequest = Json.fromJson(json, QueryRequest.class);
        } catch (Exception e) {
            return CompletableFuture.completedFuture(
                badRequest(Json.toJson(new ErrorResponse("Invalid request body", "VALIDATION_ERROR", 400)))
            );
        }
        
        long startTime = System.currentTimeMillis();
        String queryUrl = mindsdbUrl + "/api/sql/query";
        
        ObjectNode requestBody = Json.newObject();
        requestBody.put("query", queryRequest.getQuery());
        
        Duration timeout = queryRequest.getTimeout() != null 
            ? Duration.ofSeconds(queryRequest.getTimeout()) 
            : mindsdbTimeout;
        
        return wsClient.url(queryUrl)
            .addHeader("Content-Type", "application/json")
            .addHeader("Accept", "application/json")
            .setRequestTimeout(timeout)
            .post(requestBody)
            .thenApply(response -> {
                long executionTime = System.currentTimeMillis() - startTime;
                
                if (response.getStatus() == 200) {
                    JsonNode responseJson = response.asJson();
                    ArrayNode data;
                    
                    if (responseJson.isArray()) {
                        data = (ArrayNode) responseJson;
                    } else if (responseJson.isObject()) {
                        data = Json.newArray();
                        data.add(responseJson);
                    } else {
                        data = Json.newArray();
                    }
                    
                    ObjectNode metadata = Json.newObject();
                    metadata.put("query", queryRequest.getQuery());
                    metadata.put("timestamp", System.currentTimeMillis());
                    
                    QueryResponse queryResponse = new QueryResponse(
                        data,
                        data.size(),
                        executionTime,
                        metadata
                    );
                    
                    return ok(Json.toJson(queryResponse));
                } else if (response.getStatus() == 400) {
                    ObjectNode details = Json.newObject();
                    details.put("query", queryRequest.getQuery());
                    return badRequest(Json.toJson(new ErrorResponse(
                        "Invalid query syntax",
                        "QUERY_SYNTAX_ERROR",
                        400,
                        details
                    )));
                } else {
                    ObjectNode details = Json.newObject();
                    details.put("status", response.getStatus());
                    details.put("body", response.getBody());
                    return internalServerError(Json.toJson(new ErrorResponse(
                        "MindsDB query failed",
                        "MINDSDB_ERROR",
                        500,
                        details
                    )));
                }
            })
            .exceptionally(e -> {
                if (e.getCause() instanceof java.util.concurrent.TimeoutException) {
                    logger.error("MindsDB query timeout", e);
                    ObjectNode details = Json.newObject();
                    details.put("timeout", timeout.toString());
                    return status(408, Json.toJson(new ErrorResponse(
                        "Query timeout",
                        "QUERY_TIMEOUT",
                        408,
                        details
                    )));
                } else {
                    logger.error("MindsDB query error", e);
                    return internalServerError(Json.toJson(new ErrorResponse(
                        "Internal server error",
                        "SERVER_ERROR",
                        500
                    )));
                }
            });
    }
}
